### PR TITLE
TypeScript 2.7 + esModuleInterop

### LINKED ISSUE #2144: blueprint core v2.0 has an invalid es6 module build
### Bug report

- __Package version(s)__: blueprint core v2.0.0-rc.1:
- __Browser and OS versions__: all

#### Steps to reproduce

Blueprint exposes an invalid "module" field in package.json due to incorrect behaviour of TypeScript for CommonJS module imports

1. Bundle blueprint with parceljs. 
2. Since it exploses a "module" field, ~blueprint~ parceljs will use babel to compile it.
3. Since Blueprint is built with typescript, incorrect ES6 code is generated that causes the following error:

```
Uncaught TypeError: classNames is not a function
    at Spinner.render (spinner.js:24)
```

Looking at the spinner component, the ES6 generated code looks like this:

```es6
import * as tslib_1 from "tslib";
import * as classNames from "classnames";
import * as React from "react";
// ... snip ...
        var classes = classNames(Classes.SPINNER, Classes.intentClass(intent), {
// ... rest of code ...
```

Even though TypeScript allows it, importing ES5 modules as ES6 module namespaces is not allowed, and when targeting ES6 modules typescript will generate invalid code.

The solution is to import classNames using its default export:

```es6
import classNames from 'classnames'
```

or alternativey not have a "module" field in package.json so that parcel uses the main field.

#### Actual behavior

Throws above error.

#### Expected behavior

Module works correctly

### PR DESCRIPTION
**Pull request by giladgray**
#### Fixes #2144

- [ ] verify compiled files work

#### Changes proposed in this pull request:

- typescript 2.7 and enable `esModuleInterop` and `allowSyntheticDefaultImports`
- documentalist 1.1.0 for latest typedoc with 2.7 support
- tslib 1.9.0 for new helpers
  - ⚠️ dependency bump in all packages, 1.5 => 1.9

### CODE REVIEW

File: packages/core/src/components/alert/alert.tsx [hunk 1]
```diff
@@ -4,7 +4,7 @@
  * Licensed under the terms of the LICENSE file distributed with this project.
  */
 
-import * as classNames from "classnames";
+import classNames from "classnames";
 import * as React from "react";
```
Comment (adidahiya) [inline packages/core/src/components/alert/alert.tsx]: don't we need to do the same for all `import *` statements? I believe namespace imports are now discouraged altogether, so we should make that change across the board in this PR.

File: yarn.lock [hunk 2]
```diff
@@ -8326,35 +8352,35 @@ typedoc-default-themes@^0.5.0:
...
-    shelljs "^0.7.0"
+    shelljs "^0.8.1"
     typedoc-default-themes "^0.5.0"
-    typescript "2.4.1"
+    typescript "2.7.1"
 
-typescript@2.4.1:
-  version "2.4.1"
-  resolved "https://registry.yarnpkg.com/typescript/-/typescript-2.4.1.tgz#c3ccb16ddaa0b2314de031e7e6fee89e5ba346bc"
+typescript@2.7.1:
```
Comment (adidahiya) [inline yarn.lock]: can you use `yarn why` to find out why we have this constraint? I don't want to be pinned to 2.7.1 anywhere, it has bugs.

**Comment by adidahiya** [comment]
does this work with consumers using older TS versions? do they also have to enable the `esModuleInterop` to consume our typings? if they do, that's probably ok, it's a small break -- I'm sure most people are using named imports rather than `import *`

**Comment by adidahiya** [comment]
@ericanderson IIRC you're using rollup on some projects? Mind taking a look at this PR?

File: yarn.lock (hunk 2 above)
Comment (adidahiya) [reply yarn.lock]: womp. I filed https://github.com/TypeStrong/typedoc/issues/720

File: packages/core/src/components/alert/alert.tsx (hunk 1 above)
Comment (ericanderson) [reply packages/core/src/components/alert/alert.tsx]: This only needed for things that are effectively “module.exports = something” and don’t have flags marking them as originally es modules. 

TS has been doing synthetic wrong. Now you’re doing it right! (Namespaces are not invocable)

**Comment by ericanderson** [comment]
The JavaScript generated from doing this is usable by webpack, rollup, and node. What’s funny is for rollup we actually had to write plugins to convert es code from “import * as X from “Y”” to drop the star as. The es modules you were publishing before were actually wrong and are now correct

File: packages/core/src/components/alert/alert.tsx (hunk 1 above)
Comment (giladgray) [reply packages/core/src/components/alert/alert.tsx]: seems the only offenders were `classnames`, `moment`, `moment-timezone`

File: yarn.lock (hunk 2 above)
Comment (giladgray) [reply yarn.lock]: 👍 they're slightly different requests

Comment (adidahiya) [reply yarn.lock]: ok... given the response on my issue, I think we'll have to live with this.

**Comment by adidahiya** [comment]
> The es modules you were publishing before were actually wrong and are now correct

yes, that's the linked issue #2144 we're trying to fix

File: packages/core/src/components/alert/alert.tsx (hunk 1 above)
Comment (adidahiya) [reply packages/core/src/components/alert/alert.tsx]: cool, so we're still blocking this PR on fixing `import * as moment` right?

Comment (giladgray) [reply packages/core/src/components/alert/alert.tsx]: no i did that

File: packages/datetime/package.json [hunk 3]
```diff
@@ -33,7 +33,7 @@
     "@blueprintjs/core": "^2.0.0-rc.2",
     "classnames": "^2.2",
     "react-day-picker": "^7.0.7",
-    "tslib": "^1.5.0"
+    "tslib": "^1.9.0"
```
Comment (adidahiya) [inline packages/datetime/package.json]: I think all of these need to be `~1.9.0`, similar to the typescript version range

Comment (giladgray) [reply packages/datetime/package.json]: i don't think so, it's really a minimum for the new functions added to support `esModuleInterop`

Comment (adidahiya) [reply packages/datetime/package.json]: Hmm, I thought that the behavior of these functions can change across minor versions... but that doesn't seem to be the case looking at its README. We do it for internal libraries, but this should be fine; we should probably switch over to a `^` range

Comment (giladgray) [reply packages/datetime/package.json]: switch what over to `^` range?

makes sense to leave typescript with `~` to allow patches but exclude minors (cuz those can change things significantly)

Comment (adidahiya) [reply packages/datetime/package.json]: sorry, I meant switch our internal libraries which use `"tslib": "~` over to `^`. unrelated to this PR.

File: yarn.lock (hunk 2 above)
Comment (giladgray) [reply yarn.lock]: yeah for now. on my issue he said he'd have something next week for 2.7.2

Comment (aciccarello) [reply yarn.lock]: Hey guys, I've released a patch update to TypeDoc which updates Typescript to 2.7.2

**Comment by adidahiya** [comment]
I'll take this over and finish it up, hoping to publish a 2.0.0-rc.3 with this

**Pull request merged**